src/sync.cpp

```diff
@@ -2172,51 +2182,186 @@ error UnifiedSync::startSync(MegaClient* client, const char* debris, LocalPath*
 
 void UnifiedSync::changedConfigState(bool notifyApp)
 {
-    if ((mConfig.mError != mConfig.mKnownError) ||
-        (mConfig.getEnabled() != mConfig.mKnownEnabled))
+    if (mConfig.errorOrEnabledChanged())
     {
-        LOG_debug << "Sync enabled/error changing. from " << mConfig.mKnownEnabled << "/" << mConfig.mKnownError << " to "  << mConfig.getEnabled() << "/" << mConfig.mError;
+        LOG_debug << "Sync " << toHandle(mConfig.mBackupId) << " enabled/error changed to " << mConfig.mEnabled << "/" << mConfig.mError;
 
         mClient.syncs.saveSyncConfig(mConfig);
         if (notifyApp)
         {
             mClient.app->syncupdate_stateconfig(mConfig.getBackupId());
         }
         mClient.abortbackoff(false);
-
-        mConfig.mKnownError = mConfig.mError;
-        mConfig.mKnownEnabled = mConfig.getEnabled();
     }
 }
 
 Syncs::Syncs(MegaClient& mc)
-    : mClient(mc)
+  : mClient(mc)
 {
     mHeartBeatMonitor.reset(new BackupMonitor(&mClient));
 }
 
+JSONSyncConfigDB* Syncs::syncConfigDB()
+{
+    // Have we already created the database?
+    if (mSyncConfigDB)
+    {
+        // Yep, return a reference to the caller.
+        return mSyncConfigDB.get();
+    }
+
+    // Is the client using a database?
+    if (!mClient.dbaccess)
+    {
+        // Nope and we need it for the configuration path.
+        return nullptr;
+    }
+
+    // Can we get our hands on an IO context?
+    if (!syncConfigIOContext())
+    {
+        // We need it if we want to write the DB to disk.
+        return nullptr;
+    }
+
+    // Where the database will be stored.
+    auto dbPath = mClient.dbaccess->rootPath();
+
+    // Create the database.
+    mSyncConfigDB.reset(new JSONSyncConfigDB(dbPath));
+
+    return mSyncConfigDB.get();
+}
+
+bool Syncs::syncConfigDBDirty()
+{
+    return mSyncConfigDB && mSyncConfigDB->dirty();
+}
+
+error Syncs::syncConfigDBFlush()
+{
+    if (!syncConfigDBDirty())
+    {
+        return API_OK;
+    }
+
+    LOG_verbose << "Attempting to flush internal config database.";
+
+    auto result = mSyncConfigDB->write(*mSyncConfigIOContext);
```

Comment:
The direct access here was an optimization of sorts: syncConfigDBDirty() will return false if there is no sync config database. There can be no sync config database without an IO context.

It would be friendlier to future maintainers to use the acessor here, too.